libp2p/protocols/connectivity/autonat/server.nim

```diff
@@ -60,21 +60,34 @@ proc sendResponseOk(conn: Connection, ma: MultiAddress) {.async.} =
 
 proc tryDial(autonat: Autonat, conn: Connection, addrs: seq[MultiAddress]) {.async.} =
   await autonat.sem.acquire()
+  var futs: seq[Future[Opt[MultiAddress]]]
   try:
     # This is to bypass the per peer max connections limit
     let outgoingConnection = autonat.switch.connManager.expectConnection(conn.peerId)
     # Safer to always try to cancel cause we aren't sure if the connection was established
     defer: outgoingConnection.cancel()
-    # This is to bypass the global max connections limit
-    let ma = await autonat.switch.dialer.tryDial(conn.peerId, addrs).wait(autonat.dialTimeout)
+    # tryDial is to bypass the global max connections limit
+    futs = addrs.mapIt(autonat.switch.dialer.tryDial(conn.peerId, @[it]))
+    let fut = await anyCompleted(futs).wait(autonat.dialTimeout)
+    let ma = await fut
     if ma.isSome:
       await conn.sendResponseOk(ma.get())
     else:
       await conn.sendResponseError(DialError, "Missing observed address")
   except CancelledError as exc:
     raise exc
+  except AllFuturesFailedError as exc:
+    for f in futs:
+      if not f.finished():
+        f.cancel()
```

Comment:
This should happen in finally (or ideally, in finally and just after the anyCompleted)

Reply:
Why finally? But it's in the wrong place indeed. I meant to add it in `except AsyncTimeoutError`.

Reply:
You are right, we also should do that when one future completes.